src/scripting/lua_widget.cpp

```diff
@@ -0,0 +1,205 @@
+/*
+   Copyright (C) 2009 - 2018 by Guillaume Melquiond <guillaume.melquiond@gmail.com>
+   Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY.
+
+   See the COPYING file for more details.
+*/
+
+#include "scripting/lua_widget.hpp"
+#include "scripting/lua_widget_attributes.hpp"
+
+#include "log.hpp"
+#include "gui/widgets/widget.hpp"
+#include "scripting/lua_common.hpp"
+#include "scripting/lua_unit.hpp"
+#include "scripting/lua_unit_type.hpp"
+#include "scripting/lua_ptr.hpp"
+#include "scripting/push_check.hpp"
+
+
+#include "lua/lauxlib.h"
+#include "lua/lua.h"                    // for lua_State, lua_settop, etc
+
+#include <type_traits>
+
+static const char widgetKey[] = "widget";
+static char widgetdataKey[] = "widgetdata";
+
+void luaW_pushwidget(lua_State* L, gui2::widget& w)
+{
+	new(L) lua_ptr<gui2::widget>(w);
+	luaL_setmetatable(L, widgetKey);
+}
+
+gui2::widget& luaW_checkwidget(lua_State* L, int n)
+{
+	lua_ptr<gui2::widget>& lp =  *static_cast<lua_ptr<gui2::widget>*>(luaL_checkudata(L, n, widgetKey));
+	auto ptr = lp.get_ptr();
+	if(!ptr) {
+		luaL_argerror(L, n, "widget was deleted");
+	}
+	return *ptr;
+}
+
+lua_ptr<gui2::widget>& luaW_checkwidget_ptr(lua_State* L, int n)
+{
+	lua_ptr<gui2::widget>& lp =  *static_cast<lua_ptr<gui2::widget>*>(luaL_checkudata(L, n, widgetKey));
+	auto ptr = lp.get_ptr();
+	if(!ptr) {
+		luaL_argerror(L, n, "widget was deleted");
+	}
+	return lp;
+}
+
+
+bool luaW_iswidget(lua_State* L, int index)
+{
+	return luaL_testudata(L, index, widgetKey) != nullptr;
+}
+
+
+static void luaW_pushwidgettablecontainer(lua_State* L)
```

Comment:
These next few functions are kinda confusing. Could we have some sort of explanation of what they're doing, maybe some code comments?

Reply:
They are the implementation of the callbacks, basically they associate to each (c++) widget a lua table that contains their callbacl functions

Reply:
Ah, okay... I think the functions are just poorly named, then...

Reply:
well "luaW_pushwidgettable" pushed the table that is associated to a certain widget, inernally the data structure is 
{-- all windows  { --windowtable { --widget table} }} so the window table contains all widgettables for this window. and for each of these tables there are helper function that push them to the lua stack